mapeditor/inspector/townevent.cpp

```diff
@@ -0,0 +1,320 @@
+		int val = resourcesMap.value(name).toInt();
+		ui->resourcesTable->setItem(i, 0, new QTableWidgetItem(name));
+
+		QSpinBox * edit = new QSpinBox(ui->resourcesTable);
+		edit->setMaximum(i == GameResID::GOLD ? 999999 : 999);
+		edit->setMinimum(i == GameResID::GOLD ? -999999 : -999);
+		edit->setSingleStep(i == GameResID::GOLD ? 100 : 1);
+		edit->setValue(val);
+
+		ui->resourcesTable->setCellWidget(i, 1, edit);
+	}
+}
+
+void TownEvent::initBuildings()
+{
+	auto * ctown = town.town;
+	if (!ctown)
+		ctown = VLC->townh->randomTown;
+	if (!ctown)
+		throw std::runtime_error("No Town defined for type selected");
+	auto allBuildings = ctown->getAllBuildings();
+	while (!allBuildings.empty())
+	{
+		addBuilding(*ctown, *allBuildings.begin(), allBuildings);
+	}
+	ui->buildingsTree->resizeColumnToContents(0);
+
+	connect(&buildingsModel, &QStandardItemModel::itemChanged, this, &TownEvent::onItemChanged);
+}
+
+QStandardItem * TownEvent::addBuilding(const CTown& ctown, BuildingID buildingId, std::set<si32>& remaining)
+{
+	auto bId = buildingId.num;
+	const CBuilding * building = ctown.buildings.at(buildingId);
+	if (!building)
+	{
+		remaining.erase(bId);
+		return nullptr;
+	}
+
+	QString name = tr(building->getNameTranslated().c_str());
+
+	if (name.isEmpty())
+		name = QString::fromStdString(defaultBuildingIdConversion(buildingId));
+
+	QList<QStandardItem *> checks;
+
+	checks << new QStandardItem(name);
+	checks.back()->setData(bId, Qt::UserRole);
+
+	checks << new QStandardItem;
+	checks.back()->setCheckable(true);
+	checks.back()->setCheckState(params["buildings"].toList().contains(bId) ? Qt::Checked : Qt::Unchecked);
+	checks.back()->setData(bId, Qt::UserRole);
+
+	if (building->getBase() == buildingId)
+	{
+		buildingsModel.appendRow(checks);
+	}
+	else
+	{
+		QStandardItem * parent = nullptr;
+		std::vector<QModelIndex> stack;
+		stack.push_back(QModelIndex());
+		while (!parent && !stack.empty())
+		{
+			auto pindex = stack.back();
+			stack.pop_back();
+			for (int i = 0; i < buildingsModel.rowCount(pindex); ++i)
+			{
+				QModelIndex index = buildingsModel.index(i, 0, pindex);
+				if (building->upgrade.getNum() == buildingsModel.itemFromIndex(index)->data(Qt::UserRole).toInt())
+				{
+					parent = buildingsModel.itemFromIndex(index);
+					break;
+				}
+				if (buildingsModel.hasChildren(index))
+					stack.push_back(index);
+			}
+		}
+
+		if (!parent)
+			parent = addBuilding(ctown, building->upgrade.getNum(), remaining);
+
+		if (!parent)
+		{
+			remaining.erase(bId);
+			return nullptr;
+		}
+
+		parent->appendRow(checks);
+	}
+
+	remaining.erase(bId);
+	return checks.front();
+}
+
+void TownEvent::initCreatures()
+{
+	auto creatures = params.value("creatures").toList();
+	auto * ctown = town.town;
+	for (int i = 0; i < 7; ++i)
+	{
+		QString creatureNames;
+		if (!ctown)
+		{
+			creatureNames.append(QString("Creature %1 / Creature %1 Upgrade").arg(i + 1));
```

Comment:
why don't you want to make this string translatable?

Reply:
Already changed, somehow you reviewed older commit?

Reply:
yeah, was looking commit by commit because it's easier to understand separate changes like that, not all changes at once